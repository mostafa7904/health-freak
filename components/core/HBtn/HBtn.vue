<template>
  <component :is="computedTag" v-bind="computedAttrs" @click="click">
    <span v-if="!loading" class="h-btn__content">
      <slot />
    </span>
    <slot v-else name="loader" />
  </component>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import {
  genBackgroundColor,
  genBorderColor,
  genBoxShadowColor,
  genTextColor
} from '~/utils/theme'

export default defineComponent({
  name: 'HealthFreakButton',

  props: {
    activeClass: {
      type: String,
      default: ''
    },
    block: Boolean,
    flat: Boolean,
    icon: Boolean,
    loading: Boolean,
    outlined: Boolean,
    plain: Boolean,
    rounded: Boolean,
    disabled: Boolean,
    tag: {
      type: String,
      default: 'button'
    },
    text: Boolean,
    tile: Boolean,
    to: {
      type: String,
      default: null,
      validator(v: string) {
        if (!v) return true
        const pattern = new RegExp('(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*')
        return !!pattern.test(v)
      }
    },
    href: {
      type: String,
      default: null
    },
    color: {
      type: String,
      default: null
    }
  },

  computed: {
    computedClasses(): any {
      return {
        'h-btn': true,
        // "h-btn--block": this.block,
        'h-btn--disabled': this.disabled,
        'h-btn--icon': this.icon,
        'h-btn--loading': this.loading,
        'h-btn--outlined': this.outlined,
        'h-btn--plain': this.plain,
        'h-btn--rounded': this.rounded,
        'h-btn--link': this.to || this.href,
        'h-btn--text': this.text,
        'h-btn--tile': this.tile
      }
    },
    computedStyles(): any {
      let styles: any = {
        ...genTextColor(this.color)
      }
      if (this.hasBg) styles = { ...styles, ...genBackgroundColor(this.color) }
      if (!this.flat && this.hasBg)
        styles = { ...styles, ...genBoxShadowColor(this.color) }
      if (this.outlined) styles = { ...styles, ...genBorderColor(this.color) }
      return styles
    },
    computedTag() {
      if (this.to) return 'nuxt-link'
      if (this.href) return 'a'
      return this.tag
    },
    computedAttrs() {
      const attrs = {
        href: this.href,
        to: this.to,
        disabled: this.disabled,
        'aria-disabled': this.disabled,
        style: { ...this.computedStyles },
        class: { ...this.computedClasses },
        ...this.$attrs
      } as any

      return attrs
    },
    hasBg(): boolean {
      return !this.text && !this.plain && !this.outlined && !this.icon
    }
  },

  methods: {
    click(e: MouseEvent): void {
      this.$emit('click', e)
    }
  }
})
</script>

<style lang="scss">
@import '~/assets/variables.scss';

.h-btn {
  position: relative;
  width: fit-content;
  align-items: center;
  display: inline-flex;
  flex: 0 0 auto;
  font-weight: 500;
  letter-spacing: -0.0027rem;
  justify-content: center;
  outline: 0;
  text-decoration: none;
  text-indent: -0.0027rem;
  text-transform: none;
  vertical-align: middle;
  white-space: nowrap;
  border-radius: map-get($map: $h-radius, $key: 'md');
  box-sizing: border-box;
  background: transparent;
  overflow: hidden;
  cursor: pointer;
  transition-duration: 0.28s;
  transition-property: box-shadow, transform, opacity;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  user-select: none;
  &__content {
    padding: 10px 12px;
    width: 100%;
    align-items: center;
    color: inherit;
    display: flex;
    flex: 1 0 auto;
    justify-content: inherit;
    line-height: normal;
    position: relative;
    transition: inherit;
    transition-property: opacity;
  }
  &--link {
    &:hover {
      text-decoration: none;
    }
  }
  &--icon > &__content {
    padding: 5px !important;
  }
  &--outlined {
    border: thin solid;
  }
  &--disabled {
    cursor: not-allowed;
    pointer-events: none;
    opacity: 0.5;
  }
  &--rounded {
    border-radius: map-get($map: $h-radius, $key: 'xl');
  }
}
</style>
