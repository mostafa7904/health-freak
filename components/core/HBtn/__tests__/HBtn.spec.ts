// Libraries
import Vue from "vue";

// Plugins
import Router from "vue-router";

// Components
import HBtn from "../HBtn.vue";

// Utilities
import {
  createLocalVue,
  mount,
  Wrapper,
  RouterLinkStub,
} from "@vue/test-utils";

describe("HBtn.vue", () => {
  let mountFunction: (options?: object) => Wrapper<Vue>;
  let router: Router;
  let localVue: typeof Vue;

  beforeEach(() => {
    router = new Router();
    localVue = createLocalVue();
    localVue.use(Router);

    mountFunction = (options = {}) => {
      return mount(HBtn, {
        localVue,
        router,
        stubs: {
          Nuxt: true,
          NuxtLink: RouterLinkStub,
        },
        ...options,
      });
    };
  });

  it("should render component and match snapshot", () => {
    const html = mountFunction().html();
    expect(html).toMatchSnapshot();
  });

  it("should render component with color prop and match snapshot", () => {
    expect(
      mountFunction({
        propsData: {
          color: "green darken-1",
        },
      }).html()
    ).toMatchSnapshot();
  });

  it("should render an <a> tag when using href prop", () => {
    const wrapper = mountFunction({
      propsData: {
        href: "http://www.google.com",
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render specified tag when using tag prop", () => {
    const wrapper = mountFunction({
      propsData: {
        tag: "div",
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should emit a click event", async () => {
    const wrapper = mountFunction();

    const click = jest.fn();
    wrapper.vm.$on("click", click);
    wrapper.trigger("click");

    await wrapper.vm.$nextTick();

    expect(click).toHaveBeenCalled();
  });

  it("should render an <a> tag when provided a to prop", () => {
    const wrapper = mountFunction({
      propsData: {
        to: "/foo",
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
