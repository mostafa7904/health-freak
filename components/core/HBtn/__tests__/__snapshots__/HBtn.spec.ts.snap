// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HBtn.vue should render an <a> tag when provided a to prop 1`] = `"<a class=\\"h-btn h-btn--link\\"><span class=\\"h-btn__content\\"></span></a>"`;

exports[`HBtn.vue should render an <a> tag when using href prop 1`] = `"<a href=\\"http://www.google.com\\" class=\\"h-btn h-btn--link\\"><span class=\\"h-btn__content\\"></span></a>"`;

exports[`HBtn.vue should render component and match snapshot 1`] = `"<button class=\\"h-btn\\"><span class=\\"h-btn__content\\"></span></button>"`;

exports[`HBtn.vue should render component with color prop and match snapshot 1`] = `"<button class=\\"h-btn\\" style=\\"color: rgb(255, 255, 255); caret-color: #43a047; background-color: rgb(67, 160, 71); border-color: #43a047; box-shadow: 0 10px 20px -10px #43a047;\\"><span class=\\"h-btn__content\\"></span></button>"`;

exports[`HBtn.vue should render specified tag when using tag prop 1`] = `"<div class=\\"h-btn\\"><span class=\\"h-btn__content\\"></span></div>"`;

exports[`HBtn.vue should render with the correct class when passed the icon prop 1`] = `"<button class=\\"h-btn h-btn--icon\\"><span class=\\"h-btn__content\\"></span></button>"`;
